core/src/main/java/com/segment/analytics/kotlin/core/Telemetry.kt

```diff
@@ -0,0 +1,272 @@
+import kotlin.math.min
+import kotlin.math.roundToInt
+
+class MetricsRequestFactory : RequestFactory() {
+    override fun upload(apiHost: String): HttpURLConnection {
+        val connection: HttpURLConnection = openConnection("https://${apiHost}/m")
+        connection.setRequestProperty("Content-Type", "text/plain")
+        connection.doOutput = true
+        connection.setChunkedStreamingMode(0)
+        return connection
+    }
+}
+
+@Serializable
+data class RemoteMetric(
+    val type: String,
+    val metric: String,
+    var value: Int,
+    val tags: Map<String, String>,
+    val log: Map<String, String>? = null,
+)
+
+fun logError(err: Throwable) {
+    println("Error sending segment performance metrics $err")
+}
+
+object Telemetry {
+    private var _enable: Boolean = true
+    private var _host: String = Constants.DEFAULT_API_HOST
+    private var _sampleRate: Double = 1.0
+    private var _flushTimer: Int = 30 * 1000 // 30s
+    private var _maxQueueSize: Int = 20
+    private var MAX_QUEUE_BYTES = 28000
+    private var _maxQueueBytes: Int = MAX_QUEUE_BYTES
+    private var _httpClient: HTTPClient = HTTPClient("", MetricsRequestFactory())
+    private var _sendWriteKeyOnError: Boolean = true
+    private var _sendErrorLogData: Boolean = false
+    private var _errorHandler: ((Throwable) -> Unit)? = null
+    private var _rateLimitTimer: Int = 0
+
+    var enable: Boolean
+        get() = _enable
+        set(value) {
+            _enable = value
+        }
+    var sendWriteKeyOnError: Boolean
+        get() = _sendWriteKeyOnError
+        set(value) {
+            _sendWriteKeyOnError = value
+        }
+
+    var sendErrorLogData: Boolean
+        get() = _sendErrorLogData
+        set(value) {
+            _sendErrorLogData = value
+        }
+    var errorHandler: ((Throwable) -> Unit)?
+        get() = _errorHandler
+        set(value) {
+            _errorHandler = value
+        }
+
+    var host: String
+        get() = _host
+        set(value) {
+            _host = value
+        }
+
+    var sampleRate: Double
+        get() = _sampleRate
+        set(value) {
+            _sampleRate = value
+        }
+
+    var flushTimer: Int
+        get() = _flushTimer
+        set(value) {
+            _flushTimer = value
+        }
+
+    var maxQueueSize: Int
+        get() = _maxQueueSize
+        set(value) {
+            _maxQueueSize = value
+        }
+
+    var maxQueueBytes: Int
+        get() = _maxQueueBytes
+        set(value) {
+            _maxQueueBytes = min(value, MAX_QUEUE_BYTES)
+        }
+
+    var httpClient: HTTPClient
+        get() = _httpClient
+        set(value) {
+            _httpClient = value
+        }
+
+    private val queue = mutableListOf<RemoteMetric>()
+    private var queueBytes = 0
+    private var queueSizeExceeded = false
+    private val seenErrors = mutableMapOf<String, Int>()
+    private var started = false
+
+    init {
+        CoroutineScope(Dispatchers.Default).launch {
```

Comment:
we should use our own scope and dispatcher to avoid leaking exceptions to app's scope.

Reply:
Hrm.  Telemetry will be around before Analytics, and Analytics' CoroutineExceptionHandler actually seems like a good place for a Telemetry.error call... which seems incompatible with reusing that scope.  Does making a new scope + handler  sound good?

Reply:
Well, now that this happens in start, maybe?  I'll take a look.

Reply:
After some sleep - not sure why I assumed it was meant to reuse an existing scope.  👍